Add LCD Support on NXP MCXN947 over FlexIO

mmahadevan108 opened the pull request:
This PR enables FlexIO support and adds support for the LCD-PAR-S035 NXP display over the FlexIO.

Depends on https://github.com/zephyrproject-rtos/zephyr/pull/70583 for some updates to the MIPI DBI implementation

mmahadevan108 committed:
include: mipi_dbi: Add defines for MIPI Type A and B

Add defines for MIPI DBI Type A based on Motorola 6800
and Type B baedon Intel 8080 bus

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: mipi_dbi: Add controller driver for NXP FlexIO LCD

Add a driver to support the NXP FlexIO LCD controller that uses
8080/6800 bus protocol.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
boards: shield: Add a shield for the NXP LCD-PAR-S035 LCD module

NXP LCD-PAR-S035 LCD module will be used by FRDM-X evaluation
kits with a parallel LCD connector.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: clock_control: Update NXP LPC Syscon driver to add FlexIO

Add support to get FlexIO clock

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
cc @jasonNXP

danieldegrasse commented:
> Depends on https://github.com/zephyrproject-rtos/zephyr/pull/70583 for some updates to the MIPI DBI implementation

@mmahadevan108 As requested, I've split the shared commits from that PR into a new one: https://github.com/zephyrproject-rtos/zephyr/pull/71454

danieldegrasse commented on boards/nxp/frdm_mcxn947/CMakeLists.txt:
```diff
@@ -6,3 +6,10 @@
 zephyr_library_sources(board.c)
+
+# Data bus width is used by the SDK driver and processes it as a compile time option
+if(CONFIG_MIPI_DBI_NXP_FLEXIO_LCDIF)
```
Can we do this processing in the CMake lists for the driver? I assume this would be needed by targets other than the FRDM board

mmahadevan108 replied:
Sure, I can move it and use the label `zephyr_flexio_lcd` that is defined at the board level and is used by the board shields. Would that work?

danieldegrasse commented on boards/shields/lcd_par_s035/doc/index.rst:
```diff
@@ -0,0 +1,39 @@
+Requirements
+************
+
+This shield can only be used with FRDM-X evaluation kits with a parallel LCD
```
Perhaps note that this is the only interface *currently enabled*? Mentioning this since the shield also has a SPI connector.

mmahadevan108 replied:
I have added some verbiage about this.

danieldegrasse commented on boards/shields/lcd_par_s035/lcd_par_s035_8080.overlay:
```diff
@@ -0,0 +1,40 @@
+	st7796s: st7796s@0 {
+		compatible = "sitronix,st7796s";
+		reg = <0>;
+		mipi-max-frequency = <43000000>;
```
Is the clock frequency of this display 43MHz? From the comment in the board DTS, it seems the display clock should be set to 1MHz, since the actual clock line for the interface is running at 1 MHz (given 16 data lines and a baud rate of 16 Mbps)

mmahadevan108 replied:
I have deleted this as this is not used by the driver

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_flexio_lcdif.c:
```diff
@@ -0,0 +1,488 @@
+	const struct mipi_dbi_config *active_cfg;
+};
+
+static int flexio_lcdif_isr(void *user_data)
```
Can we omit this ISR? It seems the parent flexio driver will skip calling the ISR callback if it is `NULL`

mmahadevan108 replied:
Thanks for catching this.

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_flexio_lcdif.c:
```diff
@@ -0,0 +1,488 @@
+	FLEXIO_MCULCD_ClearSingleBeatWriteConfig(base);
+}
+
+static int mipi_dbi_flexio_lcdif_configure(const struct device *dev,
```
We should also read the `frequency` property from the `dbi_config` structure here, and use that value times the bus width to determine the target baudrate of the FlexIO, instead of hardcoding it into devicetree.

mmahadevan108 replied:
Thanks, I will make this change.

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_flexio_lcdif.c:
```diff
@@ -0,0 +1,488 @@
+		return err;
+	}
+
+	k_usleep(delay);
```
By the API definition, this delay should be in milliseconds (`k_msleep`) not microseconds (`k_usleep`)

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_flexio_lcdif.c:
```diff
@@ -0,0 +1,488 @@
+/* The width is read from device tree by the CMake file and passed in as a
+ * compile time define
+ */
+#if (8UL == FLEXIO_MCULCD_DATA_BUS_WIDTH)
```
Just want to check here- is there *any* way we can avoid this and still use the SDK driver? Doing this will effectively make the driver single instance.

mmahadevan108 replied:
I have deleted this as we have a property for `bus_width` which can be used. However with the compile time define needed for the SDK driver, we still have the limitation of do being able to support multiple instances with different bus widths. I will make a change request to the SDK driver to make this a config property instead of a compile time define.

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_flexio_lcdif.c:
```diff
@@ -0,0 +1,488 @@
+		.flexio_dev = DEVICE_DT_GET(DT_INST_PARENT(n)),				\
+		.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),				\
+		.child = &lcdif_child_##n,						\
+		.baudrate_bps = DT_INST_PROP(n, baudrate_bps),				\
```
See comment above- I think we should calculate the baudrate within the `configure` function above, based on the display's requested frequency.

mmahadevan108 replied:
I have deleted this property

danieldegrasse commented on dts/arm/nxp/nxp_mcxn94x_common.dtsi:
```diff
@@ -618,6 +618,22 @@
+			compatible = "nxp,mipi-dbi-flexio-lcdif";
+			status = "disabled";
+		};
+		flexio0_spi0: flexio0_spi0 {
```
Why add the SPI node? Is this needed for this PR? I mainly ask since we don't seem to enable this anywhere, so this isn't really being tested

mmahadevan108 replied:
I chose to add the flexio-spi compatible so that we are aware this support exists and it can be enabled in future.

danieldegrasse replied:
I'd prefer we add this in a separate PR- generally if we define a node with a given compatible in an SOC devicetree, I think the best practice is to have at least one test in tree that is set up to verify that the device defined by the node actually functions. So if we define the flexio_spi node, we should also add an overlay to the spi loopback test to enable testing this functionality

mmahadevan108 commented:
cc @zejiang0jason

mmahadevan108 committed:
boards: frdm_mcxn947: Add Display support

Add support for the LCD-PAR-S035 display over the
FlexIO interface.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

danieldegrasse commented on include/zephyr/dt-bindings/mipi_dbi/mipi_dbi.h:
```diff
@@ -53,6 +53,14 @@
  */
 #define MIPI_DBI_MODE_SPI_4WIRE 0x2
+/**
+ * Parallel Bus protocol for MIPI DBI Type A based on Motorola 6800 bus.
```
Can you add signal diagrams for these like we have for 3 and 4 wire mode? Perhaps we can condense the parallel wire diagram like so:
```
 *               -.-------.--------.
 *     DOUT[8:0]  | D[8:0]| D[16:8]|
 *               -'-------'--------'
 * 
 ```

danieldegrasse replied:
I should be more clear here- I'd like the diagram to also include signals like `CSX`, `RDX`, and `WRX`. The complete diagrams should look *similar* to the following from the MIPI DBI spec:


![image](https://github.com/zephyrproject-rtos/zephyr/assets/34665051/7af5105e-c80f-4035-8b0a-c095071ed233)

mmahadevan108 replied:
Can the use refer to the spec instead of trying to capture this here.

danieldegrasse replied:
They can and should refer for more information- the reason I'd like to provide the signalling diagrams here is that "mode b" and "mode a" are the names the MIPI spec uses, but not many display controller datasheets use those names. The signalling diagrams will make it easier for developers to match their display's interface description with one of the MIPI DBI modes.

mmahadevan108 replied:
I have added the diagram. Please check.

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_flexio_lcdif.c:
```diff
@@ -0,0 +1,471 @@
+		&mcux_flexio_lcdif_data_##n,						\
+		&mcux_flexio_lcdif_config_##n,						\
+		POST_KERNEL,								\
+		CONFIG_MIPI_DBI_INIT_PRIORITY,					\
```
Nit- align `\` here with others

danieldegrasse commented on boards/shields/lcd_par_s035/Kconfig.shield:
```diff
@@ -0,0 +1,5 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+config SHIELD_LCD_PAR_S035
```
It does not seem to me that this Kconfig would be set when building for the mcx board with this shield, IE with:
`west build -p always -b frdm_mcxn947//cpu0 samples/drivers/display -DSHIELD='lcd_par_s035_8080'`.

Maybe we should add another Kconfig to check for the `lcd_par_s035_8080` shield name?

mmahadevan108 replied:
Thanks, I have update the `Kconfig`

decsny commented on west.yml:
```diff
@@ -193,7 +193,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: cd046faf8a3dc879982ff6079fc48a80a0d53e7c
+      revision: pull/371/head
```
please just put this in the same commit as it is needed for

zejiang0jason commented:
From the samples/drivers/display's description, the rectangle color is: in clockwise order, from top left corner: red, green, blue, grey
https://github.com/zephyrproject-rtos/zephyr/blob/69c5c6d5117ff9b045e1b74d40b29d86ad1fe550/samples/drivers/display/README.rst?plain=1#L13

But the result is:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/da36d9bf-6291-4d08-a1eb-184263a75a0e)

Doesn't match.

danieldegrasse commented:
> From the samples/drivers/display's description, the rectangle color is: in clockwise order, from top left corner: red, green, blue, grey

@mmahadevan108 have you tested this enablement with any of the LVGL samples? Generally I find those useful to make sure color data is not swapped, as @zejiang0jason suggests it is

mmahadevan108 committed:
drivers: display: st7796: Update the logic to invert RGB and BGR

1. Add a property for panels where the RGB is displayed as BGR.
2. Add a check for 8080 8-bit mode and invert RGB and BGR for
   this case.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
> From the samples/drivers/display's description, the rectangle color is: in clockwise order, from top left corner: red, green, blue, grey
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/69c5c6d5117ff9b045e1b74d40b29d86ad1fe550/samples/drivers/display/README.rst?plain=1#L13
> 
> But the result is: ![image](https://private-user-images.githubusercontent.com/31212062/326315558-da36d9bf-6291-4d08-a1eb-184263a75a0e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUyNzg2MTksIm5iZiI6MTcxNTI3ODMxOSwicGF0aCI6Ii8zMTIxMjA2Mi8zMjYzMTU1NTgtZGEzNmQ5YmYtNjI5MS00ZDA4LWExZWItMTg0MjYzYTc1YTBlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA5VDE4MTE1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVkNWZhMDIxZGQ3MjJhNjY4NzczNGExNWQwNzk2ZTg4MmQ5MTkxYmRiOGY4MTcyNzMyZGE4MTZlMDMwYjNkNGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ljCC5Ll9LYcyS1vORtMuRETvl3OXGKT18DusD3wkX8g)
> 
> Doesn't match.

@zejiang0jason , @danieldegrasse, thank you for spotting this. I have added a change to the display driver.

mmahadevan108 commented:
> > From the samples/drivers/display's description, the rectangle color is: in clockwise order, from top left corner: red, green, blue, grey
> 
> @mmahadevan108 have you tested this enablement with any of the LVGL samples? Generally I find those useful to make sure color data is not swapped, as @zejiang0jason suggests it is

I have tested with `zephyr/samples/subsys/display/lvgl`

zejiang0jason commented on boards/shields/lcd_par_s035/Kconfig.defconfig:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2024 NXP
+# SPDX-License-Identifier: Apache-2.0
+
```
is it needed to add
```c
if SHIELD_LCD_PAR_S035
```
?

danieldegrasse replied:
I do think this is needed- it follows the pattern of other shields: https://github.com/zephyrproject-rtos/zephyr/blob/76559f27fd6e9219516c9ee7deebbdf5b3116105/boards/shields/adafruit_2_8_tft_touch_v2/Kconfig.defconfig#L4

mmahadevan108 commented:
@danieldegrasse @decsny, kindly re-review

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_flexio_lcdif.c:
```diff
@@ -0,0 +1,471 @@
+	int ret;
+
+	ARG_UNUSED(pixfmt);
+	ARG_UNUSED(dbi_config);
```
Can drop this line, the argument is used

danieldegrasse commented on boards/shields/lcd_par_s035/lcd_par_s035_8080.overlay:
```diff
@@ -0,0 +1,41 @@
+	};
+};
+
+&zephyr_flexio_lcd {
```
We should name the nodelabel `nxp_flexspi_lcd`- Zephyr implies this node label is generic, which it is not

groncarolo reviewed (commented):
I had a similar problem on RT595
I tried LVGL demo and got correct colors  

Using 
CONFIG_MIPI_DSI_MCUX_2L_SWAP16=y flag as they do in LVGL demo
solved my issue too

It might be worth checking if this would be enough in this case too

danieldegrasse reviewed (dismissed):
Changes look good to me- if you'd like (just a suggestion, not a requirement), adding the following input codes to the buttons on this board will also allow us to run the LVGL demos (`samples/modules/lvgl/demos`)

```diff
diff --git a/boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi b/boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi
index 3154af9d591..6ed05d512cb 100644
--- a/boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi
+++ b/boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi
@@ -6,6 +6,7 @@

 #include "frdm_mcxn947-pinctrl.dtsi"
 #include <zephyr/dt-bindings/i2c/i2c.h>
+#include <zephyr/dt-bindings/input/input-event-codes.h>

 / {
        aliases{
@@ -40,11 +41,13 @@
                user_button_2: button_0 {
                        label = "User SW2";
                        gpios = <&gpio0 23 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
+                       zephyr,code = <INPUT_KEY_0>;
                        status = "disabled";
                };
                user_button_3: button_1 {
                        label = "User SW3";
                        gpios = <&gpio0 6 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
+                       zephyr,code = <INPUT_KEY_1>;
                        status = "disabled";
                };
        };

```

mmahadevan108 commented:
> Changes look good to me- if you'd like (just a suggestion, not a requirement), adding the following input codes to the buttons on this board will also allow us to run the LVGL demos (`samples/modules/lvgl/demos`)
> 
> ```diff
> diff --git a/boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi b/boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi
> index 3154af9d591..6ed05d512cb 100644
> --- a/boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi
> +++ b/boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi
> @@ -6,6 +6,7 @@
> 
>  #include "frdm_mcxn947-pinctrl.dtsi"
>  #include <zephyr/dt-bindings/i2c/i2c.h>
> +#include <zephyr/dt-bindings/input/input-event-codes.h>
> 
>  / {
>         aliases{
> @@ -40,11 +41,13 @@
>                 user_button_2: button_0 {
>                         label = "User SW2";
>                         gpios = <&gpio0 23 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
> +                       zephyr,code = <INPUT_KEY_0>;
>                         status = "disabled";
>                 };
>                 user_button_3: button_1 {
>                         label = "User SW3";
>                         gpios = <&gpio0 6 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
> +                       zephyr,code = <INPUT_KEY_1>;
>                         status = "disabled";
>                 };
>         };
> ```

This is something @zejiang0jason is working on as part of his PR which he plans to submit once this support is merged in.

mmahadevan108 commented:
@zejiang0jason @decsny @dleach02 could you kindly re-review

danieldegrasse commented:
I'll add that I tested this PR using the `samples/drivers/display` and `samples/subsys/display/lvgl` samples and can confirm both work correctly for me as well

decsny commented on dts/bindings/display/sitronix,st7796s.yaml:
```diff
@@ -104,3 +104,12 @@ properties:
+      Inverting color format order (RGB->BGR or BGR->RGB)
+      In the case of enabling this option, API reports pixel-format in capabilities
+      as the inverted value of the RGB pixel-format specified in MADCTL.
+      This option is convenient for supporting displays with bugs
```
displays with bugs? what does this mean?

mmahadevan108 replied:
DIsplays that have the behave as BGR instead of when requesting RGB and vice versa. This is a similar to the https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/display/sitronix%2Cst7735r.yaml#L127C5-L133C75

zejiang0jason replied:
Hello @mmahadevan108 , @decsny ,

Although the description here and the comment in st7735 driver says it is display controller's bug, I have different understanding.

https://github.com/zephyrproject-rtos/zephyr/blob/4efc03520966a96e1a5dcf5369f75503626c2b42/drivers/display/display_st7735r.c#L276

Firstly, for RGB565&BGR565, the LCD controller's pixel format name is different with Zephyr:
1. The LCD controller setting ST7796S_MADCTL_BGR is to swap color channels R and B. As shown in the LCD controller's document:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/4aee7529-bb26-420a-8965-255f3a876cb4)
2. Zephyr's RGB565 & BGR565 is byte swap but not color channel swap.
PIXEL_FORMAT_RGB_565
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/3e260f69-6031-4b38-878e-632512717a4b)
PIXEL_FORMAT_BGR_565
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/9822367f-d16f-4c11-b35b-87c30e0b2d1d)
We can get this from the function get_rgb565_color, fill_buffer_rgb565, fill_buffer_bgr565 in display driver's sample:
https://github.com/zephyrproject-rtos/zephyr/blob/4efc03520966a96e1a5dcf5369f75503626c2b42/samples/drivers/display/src/main.c#L98-L141
In short, Zephyr's RGB565 and BGR565 are just ST7796S's RGB565, in different endian modes. So ST7796S_MADCTL_BGR shall always be set.

Secondly, the transfer interface byte order needs be considered. There are SPI, I8080-8bit, I8080-16bit.
These are the ST7796S required transfer formats:
I8080-16bit
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/17adf2a4-1d93-42a4-a24a-bb22b762ee02)
I8080-8bit
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/453ec02a-6bf2-4f65-aa48-514e19659b40)
SPI
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/a73d1535-1707-4ba6-b20d-1703dd5c167d)

If the pixels in memory is PIXEL_FORMAT_BGR_565
![image](https://github.com/zephyrproject-rtos/zephyr/assets/31212062/9822367f-d16f-4c11-b35b-87c30e0b2d1d)

With I8080-16bit, flexio LCD driver sends low byte to D0-D7, high byte to D8-D16, it matches the ST7796S requirement. PIXEL_FORMAT_BGR_565 can be used.
With I8080-8bit and SPI, low byte is sent first, it doesn't match the ST7796S requirement, the pixel byte need be swapped. PIXEL_FORMAT_RGB_565 shall be used.

LVGL's RGB565 is actually PIXEL_FORMAT_BGR_565, if PIXEL_FORMAT_RGB_565 is desired, LV_COLOR_16_SWAP need be set.

In summary:
| Interface | Pixel Format | LV_COLOR_16_SWAP |
| ------------- | ------------- | ------------- |
| I8080-16bit  | PIXEL_FORMAT_BGR_565  | n |
| I8080-8bit/SPI  | PIXEL_FORMAT_RGB_565  | y |

BTW, IMHO PIXEL_FORMAT_BGR_565 is not a good naming, because generally RGB/BGR is color channel swap.

danieldegrasse replied:
Hi @zejiang0jason, thank you for this excellent explanation. First off, I agree that the `PIXEL_FORMAT_BGR_565` is poor naming on Zephyr's part. I see it is as a historical choice we simply have to live with (as changing the API would likely cause breakage downstream). We could (and likely should) improve the documentation around these values in the display API though.


Based on this, it seems like we should be reporting `RGB` and `BGR` format based on the *interface* in use, not the setting of the MADCTL bit. Would you say this is correct? If so, we probably need different MIPI mode definitions for 8 bit versus 16 bit 8080 interfaces

zejiang0jason replied:
Hi @danieldegrasse , yes, I agree that the format needs base on the _interface_ used. And we need to clearly define the byte order on the interface, because some IPs are very powerful, they support different endian modes (byte swap). For example, the LCDIC (on RW612), it is 8-bit 8080, but supports sending both low-byte first and high byte first.

danieldegrasse replied:
> For example, the LCDIC (on RW612), it is 8-bit 8080, but supports sending both low-byte first and high byte first.

Sure- we support this in Zephyr currently (at least for SPI mode). It is primarily used for LVGL, where we can skip the overhead of setting `LV_COLOR_16_SWAP`. I'm not sure how we'd allow the display driver to determine that the interface was swapping bytes. We could use the `rgb-is-inverted` property in this case, so that the display reports a byte swapped color format (this way the color format could be swapped by the LCDIC or other IP)

mmahadevan108 replied:
@danieldegrasse @zejiang0jason , thank you for your help with suggestions and testing. I have update the `st7796s.c` to add checks for 8080 8-bit mode. I have also tested with 8-bit and 16-bit data bus width.

danieldegrasse replied:
Do we still need support for the `rgb-is-inverted` property given the discussion here? Or can it be dropped?

If we have a use case I think adding support for this makes sense, but otherwise I would lean towards removing it to minimize the size of the driver

mmahadevan108 replied:
This is not needed for the display we are using on the MCX boards. However might be required for other displays.

zejiang0jason commented on boards/nxp/frdm_mcxn947/frdm_mcxn947.dtsi:
```diff
@@ -186,5 +186,21 @@
+	enwr-pin = <1>;
+	rd-pin = <0>;
+	data-pin-start = <16>;
+	data-bus-width = <16>;
```
Is there plan to support 8-bit data width? if yes, then need to define a macro `FLEXIO_MCULCD_DATA_BUS_WIDTH`:
https://github.com/zephyrproject-rtos/hal_nxp/blob/abc66979c77421fb3a140ce2e4e6ea7165cdbe8f/mcux/mcux-sdk/drivers/flexio/mculcd/fsl_flexio_mculcd.h#L44C1-L49C3

mmahadevan108 replied:
Yes, this is passed as a compile time MACRO via the CMake file. I have added a change request to the SDK driver to avoid using this macro as it has limitations. Please see my change request for details

danieldegrasse commented on dts/bindings/mipi-dbi/mipi-dbi-controller.yaml:
```diff
@@ -18,3 +18,13 @@ properties:
     required: true
     const: 0
+
+  data-bus-width:
```
I think rather than adding the data bus property to the display, we should define multiple MIPI modes for each data bus width. Something like the following:

```
#define MIPI_DBI_MODE_8080_BUS_16_BIT 0x4
#define MIPI_DBI_MODE_8080_BUS_9_BIT 0x5
#define MIPI_DBI_MODE_8080_BUS_8_BIT 0x6
```

I see two benefits to this:
- avoid duplication of bus width encoding in each display driver (I suspect this problem won't be unique to the ST7796S)
- allow MIPI DBI drivers to reconfigure their bus width at runtime (I'm aware that due to the SDK driver structure this won't work for FlexIO, but I suspect most hardware will support runtime reconfiguration like this)

mmahadevan108 replied:
@danieldegrasse, I have made this change.

mmahadevan108 committed:
drivers: display: Update setting the mode in the st7796s driver

MIPI mode is read from the device tree.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

danieldegrasse commented on drivers/display/display_st7796s.c:
```diff
@@ -323,7 +323,7 @@ static const struct display_driver_api st7796s_api = {
 						SPI_WORD_SET(8),		\
 						0),				\
-			.mode = MIPI_DBI_MODE_SPI_4WIRE,			\
+			.mode = DT_INST_PROP_OR(n, mipi_mode, 2),		\
```
```suggestion
			.mode = DT_INST_PROP_OR(n, mipi_mode, MIPI_DBI_MODE_SPI_4WIRE),     \
```
We can still use the definitions here, they are included from the `mipi_dbi.h` header

danieldegrasse commented on drivers/mipi_dbi/CMakeLists.txt:
```diff
@@ -5,3 +5,15 @@
+  dt_prop(data_bus_width PATH "${flexio0_lcd}" PROPERTY "mipi-mode")
+  # Values for mipi-mode property are defined inside dt-bindings/mipi_dbi/mipi_dbi.h.
+  # We pass a define to the SDK driver if we are using 8-bit mode.
+  if((data_bus_width EQUAL 8) OR (data_bus_width EQUAL 5))
```
Wouldn't `data_bus_width` being 8 indicate 16 bit mode?

mmahadevan108 replied:
This is looking at the values in the mipi_dbi.h file.

danieldegrasse replied:
Yes, but in both cases (`data_bus_width==8` or `data_bus_width==5`) we are setting the FLEXIO bus width to 8. Is this correct, or should we set it to 16 when `data_bus_width==8`?

mmahadevan108 replied:
Yes, this is accurate one for 8080 and one for 6800

danieldegrasse replied:
After some offline discussion, I understand this better now- the SDK driver will default to 16 bits if no value is set for this variable. The enum values `5` and `8` correspond to MIPI DBI 8 bit mode for 6800 and 8080. Since the SDK defaults to 16 bit mode, no `else` statement is needed here to set the definition in that case

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_nxp_flexio_lcdif.c:
```diff
@@ -0,0 +1,485 @@
+	status_t status;
+
+	/* 9-bit mode is not supported by the SDK driver */
+	if ((mipi_mode == MIPI_DBI_MODE_6800_BUS_9_BIT) ||
```
We should probably add an additional check here to see if the requested MIPI mode is 8 bit when the driver is configured for 16 bit (or vice versa for 16 bit mode). Not *sure* if we need this though, since the CMake logic should identify the correct MIPI DBI setting based on the chosen display

mmahadevan108 replied:
Thats right, the SDK driver configuration is picked from device tree.

decsny commented on dts/bindings/mipi-dbi/nxp,mipi-dbi-flexio-lcdif.yaml:
```diff
@@ -0,0 +1,63 @@
+      CS Pin
+      GPIO to drive the CS pin.
+
+  rs-gpios:
```
what does rs stand for? if it is reset then it needs to be reset-gpios

decsny commented on dts/bindings/mipi-dbi/nxp,mipi-dbi-flexio-lcdif.yaml:
```diff
@@ -0,0 +1,63 @@
+    required: true
+    description: |
+      CS Pin
+      GPIO to drive the CS pin.
```
what does CS stand for (probably chip select?) should put this in description
